webpack/stories/docs/ClientRouting.mdx

```diff
@@ -0,0 +1,97 @@
+import { Meta } from '@theforeman/stories';
+
+<Meta
+  title="Introduction|Client Routing"
+  parameters={{
+    storyWeight: 120,
+  }}
+/>
+
+
+# Client Routing
+Foreman uses `react-router` for rendering react pages without full page reload,
+
+## Core
+In order to add a new route in foreman core, please follow these steps:
+
+1. Create a folder under `/react_app/routes` directory
+2. Create an index file and import the wanted component:
+
+```js
+import React from 'react';
+import Page1 from './page1';
+import { PAGE1_URL } from './constants';
+
+export default {
+  path: PAGE1_URL,
+  render: props => <Page1 {...props} />,
+};
+```
+3. import that index in `routes.js` file:
+```js
+// Other routes...
+import Page1 from './Page1';
+import Page2 from './Page2'
+
+export const routes = [Page1, Page2];
+```
+4. Add a route that point to this page in `routes.rb` :
+```ruby
+match 'page1' => 'react#index', :via => :get
+```
+
+## Plugins
+You can use `react-router` in your plugin as well with no boilerplate.
+
+### Register global routes file
+In the plugin registeration please add:
+``` ruby
+Foreman::Plugin.register :"<plugin-name>" do
+  register_global_js_file 'routes'
+  # some code
+end
+```
+This tells foreman core to load a `routes_index.js` file.
+
+### Register routes
+Create `routes_index.js` file under `webpack` directory:
+
+```js
+import { addGlobalRoutes } from 'foremanReact/routes/ExternalRoutes';
+import routes from './routes';
+
+addGlobalRoutes('<plugin-name>', routes);
+
+```
+
+### Creating the routes
+Create a `routes.js`:
+
+```js
+import React from 'react';
+import IndexPage from './IndexTasks';
+import ShowPage from './ShowTask';
+
+const ForemanTasksRoutes = [
+  {
+    path: '/<plugin>/index',
+    exact: true,
+    render: props => <IndexPage {...props} />,
+  },
+  {
+    path: '/<plugin>/:id',
+    render: props => <ShowPage {...props} />,
+  },
+];
+
+export default ForemanTasksRoutes;
+
+```
+
+### Adding the routes in routes.rb
+The plugin needs to identify that a page should be loaded via react router.
+Doing so by updating the plugin's `routes.rb`:
+```ruby
+ match '/plugin_page' => '/react#index', :via => [:get]
```

Comment:
nit: remove slash in `react#index`

Reply:
the `/` renders the `foreman-core` react layout

Reply:
Tested and it works nicely, didn't know that.. thanks!
how should we treat permissions?
usually they are set in the controller or in the engine file - for a controller.. 
but now that it's pointing to foreman that's an issue, I created the same file ( which contains nothing) in the plugin and it works well and might solve the permissions issue.
although the foreman react_controller also renders the layout correctly: https://github.com/theforeman/foreman/blob/develop/app/controllers/react_controller.rb#L2

so we need to do the same for plugins or solve the permissions issue.
cc @tbrisker

Reply:
when a page renders within react the data is fetched via ajax requests, so permissions need special care from the UI approach, no doubt. but it isn't related just to react-router. during the host details page effort, we will try to find the best mechanism for that